### DISCUSSION TITLE
Future Plugin ideas

### DISCUSSION THREAD

**Discussion by hayes** [hayes/pothos#108] [Ideas]
## WithInput plugin

Mostly implemented, but not documented or tests. 
The API is slightly sub-optimal due to some typescript limitations, still considering if its worth it in its current state.

PR: #74 

## Apollo Federation plugin

Some prototyping and initial investigation has been done, typing things correctly is proving to be very challenging, and some limitations in typescript are making the API kinda awkward to work with.  Still something I want to land, just need to spend more time figuring out the details

## Prisma plugin

Some work has been started on an initial API https://gist.github.com/hayes/c9cdcbcd9b38991eaafdb4c2f237a148. 
I spiked out the types for this API (with some changes need to make it actually work).  There are a couple of things to figure out: Prisma and dataloader integration need some API design, and I need to figure out how to get this to work with Interfaces and Unions.  There also are some design issues with how models without a unique query can be loaded. 

This will probably be the next plugin that I actually release

## CASL or simpler auth

authScopes is great for certain auth patterns, but there might be a need for something that integrates with something like CASL that is slightly easier to use.

## Other high level ideas:
* Logging
* Metrics
* Complexity

**Comment by hayes** [comment]
This discussion is a bit stale now:

* Metrics + Tracing plugins have been created
* Prisma plugin exists and has gone through a lot of iteration
* a new authz plugin was added (although it is VERY basic)
* Complexity plugin exists
* WithInput plugin exists
* Apollo federation plugin exists

**Comment by arimgibson** [comment]
Sorry for adding onto a super stale thread but something I came across recently is Zenstack (an auth layer that extends Prisma client). If I find the time, I wanna spend a couple hours seeing if it could potentially be compatible with Pothos as is, but otherwise 
Zenstack support could be a great addition onto the Prisma plugin!

**Comment by kinglouie** [comment]
hey @hayes, I'm currently in the process of figuring out how to do authorization and share the access definitions between my pothos based api server and my frontend. I found CASL and thought it would be a nice fit for this, through searching casl + pothos i found this thread. 
Do you think it is viable to mash the pothos scope auth plugin and CASL together somehow or are there any major roadblocks ahead?
Would be really nice to get some input on this, because this is my first graphql project and I'm having a hard time understanding all the possible authorization edgecases that graphql brings to the table.

**Reply by hayes** [reply]
@m1212e has been working on something like this.  Checkout their post here: https://github.com/hayes/pothos/discussions/1414

**Reply by m1212e** [reply]
Hey @kinglouie!
I just added application level filters to [rumble](https://github.com/m1212e/rumble). They provide an easy way to register custom filters for entities and allow for implementing authz which is not coverable by injecting db queries. You might be interested in taking a look at the custom filter plugin at https://github.com/m1212e/rumble/blob/d40ef02c0be898aacbcc40a87f83917d86ea51af/lib/explicitFiltersPlugin/explicitFiltersPlugin.ts#L80 which does most of the filtering and ensures automatic application of filter functions. Feel free to ask if you have any question!
